ddtrace/contrib/vertica/patch.py

```diff
@@ -0,0 +1,265 @@
+                    },
+                    "fetchone": {
+                        "operation_name": "vertica.fetchone",
+                        "span_type": "vertica",
+                        "span_end": fetch_span_end,
+                    },
+                    "fetchall": {
+                        "operation_name": "vertica.fetchall",
+                        "span_type": "vertica",
+                        "span_end": fetch_span_end,
+                    },
+                    "nextset": {
+                        "operation_name": "vertica.nextset",
+                        "span_type": "vertica",
+                        "span_end": fetch_span_end,
+                    },
+                }
+            },
+        },
+    },
+)
+
+
+def patch():
+    global _PATCHED
+    if _PATCHED:
+        return
+
+    _install(config.vertica)
+    _PATCHED = True
+
+
+def unpatch():
+    global _PATCHED
+    if _PATCHED:
+        _uninstall(config.vertica)
+        _PATCHED = False
+
+
+def _uninstall(config):
+    for patch_class_path in config["patch"]:
+        patch_mod, _, patch_class = patch_class_path.rpartition(".")
+
+        for patch_routine in config["patch"][patch_class_path]["routines"]:
+            mod = importlib.import_module(patch_mod)
+            cls = getattr(mod, patch_class)
+            unwrap(cls, patch_routine)
+
+
+def _find_routine_config(config, instance, routine_name):
+    """Attempts to find the config for a routine based on the bases of the
+    class of the instance.
+    """
+    bases = instance.__class__.__mro__
+    for base in bases:
+        full_name = "{}.{}".format(base.__module__, base.__name__)
+        if full_name not in config["patch"]:
+            continue
+
+        config_routines = config["patch"][full_name]["routines"]
+
+        if full_name in config["patch"] and routine_name in config_routines:
+            return config_routines[routine_name]
+    return {}
+
+
+def _install(config):
+    for patch_class_path in config["patch"]:
+        patch_mod, _, patch_class = patch_class_path.rpartition(".")
+
+        def wrap_init():
+            _patch_item = patch_class_path
+            patch_class_routine = "{}.{}".format(patch_class, "__init__")
+
+            # patch the __init__ of the class with a Pin instance containing the defaults
+            @wrapt.patch_function_wrapper(patch_mod, patch_class_routine)
+            def init_wrapper(wrapped, instance, args, kwargs):
+                wrapped(*args, **kwargs)
+
+                # create and attach a pin with the defaults
+                Pin(
+                    service=config["service_name"],
+                    app=config["app"],
+                    app_type=config["app_type"],
+                    tags=config.get("tags", {}),
+                    tracer=config.get("tracer", ddtrace.tracer),
+                    _config=config["patch"][_patch_item],
+                ).onto(instance)
+
+        wrap_init()
```

Comment:
Why does this need to be a function, can we not just execute the code immediately?

Reply:
The reasoning behind writing these as inline functions is that it allowed me to quickly change around the logic without having to constantly update the signatures. That said, if we're happier with this approach we can clean these up into helper functions :smile: